enarx-srv-traits/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+//! The traits contained in this crate constitute a big state machine for
+//! lauching and managing a keep across multiple technologies. Launching
+//! a keep goes through multiple states, in this order:
+//!
+//!   1. Preparation
+//!   2. Building
+//!   3. Attesting
+//!   4. Launching
+//!   5. Running
+//!
+//! ## Preparation
+//!
+//! The purpose of the preparation state is to give the technology a view
+//! of all static binary pages and entry point before creating an actual
+//! keep. This is useful for technologies like SGX where we need to hash
+//! and sign the binary before creating the keep.
+//!
+//! A keep loader will start by creating a `Preparer` object by calling
+//! `Preparer::new()`. If the technology represented by this type is not
+//! supported by the system, it will return `None`.
+//!
+//! Once a `Preparer` is obtained, the keep manager will call `Loader::load()`
+//! to load all the pages of the static binary. Then finally, the manager
+//! will call `Preparer::next()` to specify the entry point and conclude the
+//! preparation state.
+//!
+//! ## Building
+//!
+//! The purpose of the building state is to actually create a full keep
+//! with all the pages actually loaded. The workflow is similar to the
+//! preparation state.
+//!
+//! ## Attestation
+//!
+//! The purpose of the attestation state is attest the keep to a remote
+//! entity (usually the tenant). This state proxies packets back and forth
+//! between the keep and the tenant. Note that the contents of attestation
+//! packets is defined by the specific technology in question.
+//!
+//! ## Launching
+//!
+//! The purpose of the launching state is to proxy communication between
+//! the tenant and the keep after attestation. Note that the intent of this
+//! state is to proxy integrity protected ciphertext. The manager should
+//! have no visibility into its contents.
+//!
+//! ## Running
+//!
+//! The purpose of the running state is to handle syscalls from the keep.
+
+use std::convert::TryFrom;
+use std::io::{Read, Result, Write};
+use std::num::TryFromIntError;
+use std::ops::BitAnd;
+
+use bitflags::bitflags;
+
+bitflags! {
+    /// Memory access permissions
+    ///
+    /// Raw numeric values are from the ELF specification.
+    pub struct Access: u8 {
+        /// Execute access
+        const X = 1 << 0;
+
+        /// Write access
+        const W = 1 << 1;
+
+        /// Read access
+        const R = 1 << 2;
+    }
+}
+
+/// A keep request
+pub enum Request {
+    Halt,
+}
+
+/// A keep reply
+pub enum Reply {
+    Halt,
+}
+
+/// A trait for supplying statically linked pages into the keep
+pub trait Loader {
+    fn load(&mut self, pages: &[u8], offset: u64, access: Access) -> Result<()>;
```

Comment:
What about .bss sections with `filesize = 0` and `memsize > 0`

Reply:
Good question. In SGX these have to be pre-expanded. That is, you allocate a temporary empty page and pass it in. But I suppose this could be handled internally for those technologies that require it. Is this a better API?

```suggestion
    fn load(&mut self, src: &[u8], dst: Range<u64>, access: Access) -> Result<()>;
```

Things like `src` alignment and intermediate temporary buffers need to be fixed up internally, I guess.

Reply:
I have done something similar in my latest push (forthcoming...).